hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/DiskCheckUtil.java

```diff
@@ -0,0 +1,199 @@
+  private DiskCheckUtil() { }
+
+  // For testing purposes, an alternate check implementation can be provided
+  // to inject failures.
+  private static DiskChecks impl = new DiskChecksImpl();
+
+  @VisibleForTesting
+  public static void setTestImpl(DiskChecks diskChecks) {
+    impl = diskChecks;
+  }
+
+  @VisibleForTesting
+  public static void clearTestImpl() {
+    impl = new DiskChecksImpl();
+  }
+
+  public static boolean checkExistence(File storageDir) {
+    return impl.checkExistence(storageDir);
+  }
+
+  public static boolean checkPermissions(File storageDir) {
+    return impl.checkPermissions(storageDir);
+  }
+
+  public static boolean checkReadWrite(File storageDir, File testFileDir,
+      int numBytesToWrite) {
+    return impl.checkReadWrite(storageDir, testFileDir, numBytesToWrite);
+  }
+
+  /**
+   * Defines operations that must be implemented by a class injecting
+   * failures into this class. Default implementations return true so that
+   * tests only need to override methods for the failures they want to test.
+   */
+  public interface DiskChecks {
+    default boolean checkExistence(File storageDir) {
+      return true;
+    }
+    default boolean checkPermissions(File storageDir) {
+      return true;
+    }
+    default boolean checkReadWrite(File storageDir, File testFileDir,
+                            int numBytesToWrite) {
+      return true;
+    }
+  }
+
+  /**
+   * The default implementation of DiskCheck that production code will use
+   * for disk checking.
+   */
+  private static class DiskChecksImpl implements DiskChecks {
+
+    private static final Logger LOG =
+        LoggerFactory.getLogger(DiskCheckUtil.class);
+
+    @Override
+    public boolean checkExistence(File diskDir) {
+      if (!diskDir.exists()) {
+        logError(diskDir, "Directory does not exist.");
+        return false;
+      }
+      return true;
+    }
+
+    @Override
+    public boolean checkPermissions(File storageDir) {
+      // Check all permissions on the volume. If there are multiple permission
+      // errors, count it as one failure so the admin can fix them all at once.
+      boolean permissionsCorrect = true;
+      if (!storageDir.canRead()) {
+        logError(storageDir,
+            "Datanode does not have read permission on volume.");
+        permissionsCorrect = false;
+      }
+      if (!storageDir.canWrite()) {
+        logError(storageDir,
+            "Datanode does not have write permission on volume.");
+        permissionsCorrect = false;
+      }
+      if (!storageDir.canExecute()) {
+        logError(storageDir, "Datanode does not have execute" +
+            "permission on volume.");
+        permissionsCorrect = false;
+      }
+
+      return permissionsCorrect;
+    }
+
+    @Override
+    public boolean checkReadWrite(File storageDir,
+        File testFileDir, int numBytesToWrite) {
+      File testFile = new File(testFileDir, "disk-check-" + UUID.randomUUID());
```

Comment:
disk-check-** files can be removed if already existing due to previous random failure.

Reply:
If there is a failure here, files are deleted on startup/shutdown currently. Technically we could delete the file here as well since volume checks are synchronized and we don't need to worry about deleting another checks' file. However, this util class is currently agnostic to those synchronization details higher up in the stack. I don't think its correctness should depend on the callers having to do coordination among themselves even if the code happens to be doing that right now.